### DISCUSSION TITLE
我封装了一层formily,在外层怎么调用formily的生命周期？

### DISCUSSION THREAD

**Discussion by heianqishizhisuiyuan** [alibaba/formily#2929] [Q&A 2.x] [answered]
我自己用formily封装了个通用组件，我在这个通用组件内创建了createForm，并把这个实例抛到外层，在调用这个组件是，我应该怎么优雅的（正确的）调用表单的副作用钩子（onFormInit, onFormMount,onFormUnmount等），字段的副作用钩子（onFieldMount等）
[codesandbox](https://codesandbox.io/s/magical-dew-9nvub8?file=/FormilyForm.tsx)
我有以下几点疑惑：
 1. 我这种form实例抛出到外层的方式是不是正确的使用方式（优雅的方式）
 2.  在外层调动表单副作用必须先从"@formily/core"引入相应的副作用，然后调用，有没有像formily v1版本的方式： $('onFieldInputChange', '*').subscribe(() => {  }) 直接使用，不用引入相关的副作用方法。
 3.在外层调用表单的副作用（onFormMount），我是通过回调过来的实例，然后用form.addEffects()方式去实现的，我的写法是不是正确的写法（个人感觉有些复杂）？       表单的onFormInit副作用应该怎么写？

**Comment by janryWang** [comment, marked as answer]
form放到外层实例化，传给你的组件不就行了

**Reply by heianqishizhisuiyuan** [reply]
如果放到外层实例化确实能解决我提到的问题，但是对于使用我封装组件的人就会觉得多了一个步骤，有点麻烦了，而且后续我也有打算把formily常用的api根据我们团队的开发习惯，书写习惯，在封装一层，如果把实例化放到外面就不容易实现了

**Reply by janryWang** [reply]
不要觉得多写一步麻烦，这是最佳实践

**Reply by guopingxiao** [reply]
哈哈哈，最佳实践~

**Reply by heianqishizhisuiyuan** [reply]
好的，那我就按照这种方式优化下我封装的代码，谢谢大神的指点了 -_-

**Reply by beita-wang** [reply]
我也遇到了你这个问题，封装了个通用组件，我在这个通用组件内创建了createForm，并把这个实例抛到外层。 那我在外层就没办法执行onFieldMount 或者onFieldInit 等生命周期。因为这些生命周期都要包裹在createForm effect中。 我也尝试使用addEffects来注册生命周期，但是会出现不可预测的问题，生命周期执行过的不会再重复执行。 所以想问下该怎么设计比较合理，有什么demo吗~
